### ISSUE TITLE
Please stop covertly canvas fingerprinting users

### ISSUE THREAD

**Issue by kuro68k** [dear-github/dear-github#271]
Github uses covert canvas fingerprinting on every page.

Canvas fingerprinting uses hidden HTML canvas elements to render images that can be used to fingerprint individual computers. In that way Github is able to track which machines a user is browsing the site from, at what times of the day and what they are doing with each machine.

This is extremely creepy. Some browsers have features to block canvas fingerprinting, either natively or via add-on.

Please stop using canvas fingerprinting and any other underhand tracking methods.

**Comment by hoganlong** [comment]
@kuro68k -- Do you know that Github is doing this -- or are you just expressing your hope that if they were they should stop?

**Comment by kuro68k** [comment]
![canvas fingerprints](https://i.imgur.com/WmCKYFM.png)

**Comment by hoganlong** [comment]
I agree -- particularly if cookies are enabled there is no reason to use this technique.

**Comment by hoganlong** [comment]
@kuro68k -- I installed this plugin and I don't get the same results -- which browser are you using?  Maybe firefox auto-blocks this?

**Comment by kuro68k** [comment]
Chrome. I also have uBlock Origin and Privacy Badger.

**Comment by avbentem** [comment]
> I added evidence to the first post.

I see Canvas is used, but I don't see evidence of fingerprinting there? Is the plugin doing anything smart to see if the results are actually sent back to the GitHub servers?

Obviously HTML5 Canvas has many use cases. For the icons you're showing: maybe GitHub is combining or colouring icons client side. I don't know what GitHub is trying to do, but not all usage of Canvas is fingerprinting.

(Automatically blocking it, using Brave browser, has not affected my usage of GitHub in any way, it seems. That by itself does not indicate much: it's either being used for something that's just candy and not required, or it's fingerpinting indeed. As I'm logged in to GitHub anyway I've not investigated this specific usage.)

**Comment by kuro68k** [comment]
What possible explanation could there be for rendering random characters into a hidden cavas other than fingerprinting?

**Comment by avbentem** [comment]
I cannot speak for GitHub, but for an OpenLayers project I'm involved in, we use Canvas to color icons. After applying the color, we need to read back the icons to pass them to standard OpenLayers methods that place the icons on the map, which indeed triggers the Brave fingerprinting warning.

Are you seeing different characters all the time, or the same emoticons for each alert?

**Comment by avbentem** [comment]
After installing the plugin I see the same emoticons being tried, with different results. I don't know why, but I'd guess GitHub is trying to figure out if the browser/OS supports recent emoticons, like https://emojipedia.org/couple-with-heart-woman-woman/ which on non-supporting platforms will show as three separate emoticons instead:

![image](https://user-images.githubusercontent.com/1497002/32690229-b2179e08-c6f3-11e7-90ca-25c1f227b012.png)

**Comment by avbentem** [comment]
(Just a ping as I've edited my last comment many times.)

**Comment by kuro68k** [comment]
Notice how the emoji in my image are black and white? They render differently on different computers, depending on the browser, OS, installed fonts etc. Most of the rendered characters are not used on the page anywhere. If they were being used for later display then why aren't the icons on the page black and white like the hidden canvas ones?

Also look at the underscores. Because they render off pixel boundaries the anti-aliasing creates a fairly unique fingerprint. It's a classic fingerprinting technique, no doubt about it.

**Comment by avbentem** [comment]
As for black and white: how did the following from my earlier comment show in your browser?

![image](https://user-images.githubusercontent.com/1497002/32692730-a203f0d2-c71d-11e7-88b0-be35e40022c0.png)

**Comment by kuro68k** [comment]
Full colour.

**Comment by hoganlong** [comment]
> It's a classic fingerprinting technique, no doubt about it.

It is a very new fingerprinting technique --  nothing classic about it.

**Comment by kuro68k** [comment]
This is definitely nothing to do with emoji. The attempts to generate canvases are similar on every page, and unrelated to the content.

**Comment by avbentem** [comment]
Did you already find some prove something is sent back to GitHub? And did you see what differences you get in the emoji in the above comments when you allow reading Canvas (in an incognito window if you like)?

Even when not used in any user content, about every GitHub page uses emoji in the buttons to respond to comments. Just look at the HTML to see many occurrences of `<g-emoji>`. When peeking in the JavaScript sources you'll see references to the same element, and things like `github/legacy/behaviors/g-emoji-element`, `isEmojiSupported` or `"Unable to get 2d canvas -- github/emoji-detection.js:35"`. Of course, without any further investigation all that does not prove that no fingerprinting is used. But the usage of Canvas' `getImageData` does not prove fingerprinting _is_ used either.

**Comment by kuro68k** [comment]
Try blocking Github's "octolytics" surveillance system. The canvas fingerprinting goes away. It's a product of their analytics system.

Some info here, but no mention of the canvas fingerprinting: https://johnnunemaker.com/analytics-at-github/

**Comment by avbentem** [comment]
> Try blocking Github's "octolytics" surveillance system. 

How?

**Comment by kuro68k** [comment]
uBlock Origin works quite well for that sort of thing.

**Comment by avbentem** [comment]
Just running uBlock Origin with its default settings does not stop the use of Canvas in my Chrome on a Mac. (When logged in, I get 3 requests blocked on this very page, including `api.github.com/_private/browser/stats`. But the emoji detection still runs, and is still allowed to use Canvas' `getImageData`, and hence shows me the built-in emoji characters, not some `<img>` as a fallback.)

Canvas is surely used to support all emoji in all browsers, like proven by the fallback to `<img>` as explained in my earlier comments. (And, as you're seeing full colour emoji too, it seems your browser is using the fallback too, as the built-in emoji seem to be black and white in your browser when drawn on a canvas.) 

And surely statistics about the support of emoji are being sent back home too, but I've not seen any prove of fingerprinting. But again, I've not investigated a lot.

**Comment by kuro68k** [comment]
These hidden canvases have nothing to do with emoji. That's just the random text they render as part of the fingerprinting.

If you block those hidden canvases, which some browsers support natively or via add-on, the emoji on the page still work.

**Comment by avbentem** [comment]
Yes, the emoji still work when blocking read access to Canvas, as GitHub then replaces the emoji with images. The prove of that is in the HTML, and explained in my first comments.

**Comment by kuro68k** [comment]
It makes no sense for it not to be fingerprinting. Where are the "little girl" and female gender symbols used? Why are there two of some of them? Why are there multiple underline characters, which are typically used for fingerprinting because they create unique gaps between them?

**Comment by avbentem** [comment]
> Where are the "little girl" and female gender symbols used?

Well, they are used in my examples above. Here's another one: :female_detective: 

> Why are there two of some of them?

Like I wrote above, with emphasis now:

>  I'd guess GitHub is trying to figure out if the browser/OS supports recent emoticons, like https://emojipedia.org/couple-with-heart-woman-woman/ **which on non-supporting platforms will show as three separate emoticons instead** (like shown in the screenshot in the first comment)

Follow the link if you don't believe me.

> underline characters, which are typically used for fingerprinting 

The plugin you're using shows the underlines are not even included in the strip it tries to read back. My guess is that the whitespace _above_ those underlines is used to detect how much width the actual emoji take, in order to detect if the emoji was rendered, and if it was rendered as a single combined emoji or a sequence of emoji, and hence to determine the level of emoji support. (Which you can see in the statistics.) Just guesses; I've not bother to decipher the JavaScript.

Anyway, as by not accepting that emoji are used it seems you're not bothering to read my comments, peek in the code or look at the network traffic. And as I think you actually 😍 emoji, I'm out of here.

![image](https://user-images.githubusercontent.com/1497002/33260254-8c713592-d35f-11e7-822d-bf4adfba5b1a.png)

**Comment by kuro68k** [comment]
I did bother to check the JS. That's just the first read, check the later ones which try to access other parts of the canvas.

Your example emoji prove the point: the set is the same on every page regardless of which emoji are actually visible on the page itself.

There may be additional legitimate uses for these canvases, but they have all the hallmarks of fingerprinting. In any case, as more and more people start blocking canvas fingerprinting (Firefox now includes blocking, enabled by default in Private Browsing and optional in normal browsing) it's a bad idea to rely on it for general site operation.

**Comment by mikerockett** [comment]
To think that the fingerprint blocker did the one single thing I’ve been looking for forever. Windows’ emoji’s are *terrible* (to say the least), so this allows GH to fallback to images, which is *perfect*.

**Comment by DaanBiesterbos** [comment]
I don't give a damn. There should be laws on how data is used. Considered the possibility that these technologies are also useful to detect abuse?
